Update Left-side overlay positions

cari-lynn opened the pull request:
This work is to update the position (file name and line number) of left side node when a document is overlayed. It is needed to report right side schema error messages, and also shows up when using the `--output pos` flag.

cari-lynn committed:
Add failing test for map overlay position

- need to build in a way to test line positions `template_test.go`
- run this test with `./hack/test-unit.sh -run TestYAMLTemplate
  TestYAMLTemplate.filetest=filetests/ytt-library/overlay/replace-key-position`

cari-lynn committed:
Add ability to assert positions in template_test

Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>

cari-lynn committed:
Extract converting evaluated output as string

- extract stringifying evaluated template into separate function

cari-lynn committed:
Test map output postions after overlay
Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

cari-lynn committed:
Test array ouput positions after overlay

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware commented:
We've found that in general, starlark structs have unknown positions. The unknown position now surfaces (using `--output pos`) when overlaying starlark structs. Thinking about the scope of this PR, we've decided that changes to the positions of starlark struct do not belong in this PR. If this is something the team wants to add, we should write another story to be picked up. 

Starlark struct example:
```
#@ def return_starlark():
#@   return {"foo": "something"}
#@ end
```

cc: @cari-lynn

pivotaljohn reviewed (changes requested):
Reviewed together.

Notes:
- in `template_test.go` look tearing off an interface for the `AsBytes()` function; `interface{}` means we can't commit to a given interface.
- explore where the "[Single Point Of Truth](http://www.catb.org/~esr/writings/taoup/html/ch04s02.html#spot_rule)" could be for setting a value of a target node, given a replacing node.

cari-lynn committed:
Add function to set Position of a Node

- Change return and parameter type to a more strict interface

Signed-off-by: Steven Locke <slocke@vmware.com>

cari-lynn commented:
Addressed @jtigger 's comments above.

We created a Node interface method to set the position. We did not update references to where we are setting the position directly (ie: map.Position) and did not make the field private since we noticed there is not a consistent pattern for other fields currently. Happy to hear opinions if you have any regarding it.

cppforlife commented on pkg/yamltemplate/filetests/ytt-library/overlay/array-output-position.yml:
```diff
@@ -0,0 +1,74 @@
+
++++
+
+OUTPUT POSITION:          stdin:45 | [doc]
```
should we add a newline after OUTPUT POSITION? seems like it would read way better with it.

cari-lynn replied:
Good idea 💥

cari-lynn committed:
Add newline in output position tests

cari-lynn committed:
Add comment that keeping fields is intentional

- Preserving the Type and Metas fields on merge overlay operations is
  intentional in the anticipation that we may need them later for type
checking
- We did not change logic to preserve fields on overlay replace
  operations, but anticipate we may need to preserve some fields later
on.
Signed-off-by: John Ryan <jtigger@infosysengr.com>

pivotaljohn reviewed (approved):
Reviewed synchronously.

In terms of "Single Point of Truth" — thanks for looking into this. Seems fine to let this ride for a bit as we watch the schema feature settling in. The comments left there I think will help future readers understand the current intent and what might change that implementation.
